libkbfs/folder_branch_ops.go

```diff
@@ -738,6 +764,109 @@ func (fbo *folderBranchOps) startMonitorChat(tlfName tlf.CanonicalName) {
 	})
 }
 
+func (fbo *folderBranchOps) kickOffRootBlockFetch(
+	ctx context.Context, rmd ImmutableRootMetadata) <-chan error {
+	ptr := rmd.Data().Dir.BlockPointer
+	return fbo.config.BlockOps().BlockRetriever().Request(
+		ctx, defaultOnDemandRequestPriority, rmd, ptr, NewDirBlock(),
+		TransientEntry)
+}
+
+func (fbo *folderBranchOps) waitForRootBlockFetch(
+	ctx context.Context, rmd ImmutableRootMetadata,
+	fetchChan <-chan error) error {
+	fbo.log.CDebugf(ctx, "Ensuring that the latest root directory "+
+		"block for revision %d is available", rmd.Revision())
+	select {
+	case err := <-fetchChan:
+		fbo.log.CDebugf(ctx, "Root block fetch complete: +%v", err)
+		return err
+	case <-ctx.Done():
+		return errors.WithStack(ctx.Err())
+	}
+}
+
+func (fbo *folderBranchOps) commitFlushedMD(
+	rmd ImmutableRootMetadata, updatedCh <-chan struct{}) {
+	diskMDCache := fbo.config.DiskMDCache()
+	if diskMDCache == nil {
+		return
+	}
+
+	// Bail out if the latest merged revision has already been updated.
+	select {
+	case <-updatedCh:
+		return
+	default:
+	}
+
+	ctx := fbo.ctxWithFBOID(context.Background())
```

Comment:
> Why does this make its own context instead of passing it from setHeadLocked?

Because this runs in a goroutine, so it shouldn't rely on someone else's maybe-canceled context.

> Actually maybe another approach here is for latestMergedUpdated to be a Context and CancelFunc, which is canceled and reassigned upon an update. 

I know it's controversial, but I still try to subscribe to the "no holding of contexts" suggestion.  And also in this case, it seems like doing would introduce more state to manage and not eliminate any of the `select` waits, since it's not the RPCs that are blocking; instead, we're getting back channels from those calls that need explicit waiting anyway.  So either we wait on the channel, or on `ctx.Done()`.

Reply:
All right, feel free to leave it then.